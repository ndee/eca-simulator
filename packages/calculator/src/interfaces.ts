export enum PurchaseRepeat {
  Daily = "Daily",
  Weekly = "Weekly",
  Monthly = "Monthly",
  Yearly = "Yearly",
}

export interface IEcaPurchase {
  purchaseAmountInEur: number;
  purchaseRepeat: PurchaseRepeat;
  purchaseStart: Date;
  purchaseEnd: Date;
}
export interface IEcaResult {
  acquiredSatoshis: number;
  investedEur: number;
}

export interface IPurchaseRequest {
  amountInEur: number;
  date: Date;
}

export interface IPurchase extends IPurchaseRequest {
  btcUsdRate: number;
  usdEurRate: number;
  acquiredSatoshis: number;
  exchangeRateSource?: string;
}

export enum ExchangeRateType {
  BtcToUsd = "BtcToUsd",
  UsdToEur = "UsdToEur",
}

export type ExchangeRateProvider = (
  dates: Date[],
  rateType: ExchangeRateType
) => Promise<number[]>;

export interface IBuildPurchaseParams {
  requestList: IPurchaseRequest[];
  exchangeRateProvider: ExchangeRateProvider;
}
