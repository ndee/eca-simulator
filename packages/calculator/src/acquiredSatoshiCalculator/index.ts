import { IEcaPurchase, IEcaResult } from "../interfaces";

export async function calculateAcquiredBitcoin(
  ecaPurchase: IEcaPurchase
): Promise<IEcaResult> {
  return { acquiredSatoshis: 1234, investedEur: 12 } as IEcaResult;
}
