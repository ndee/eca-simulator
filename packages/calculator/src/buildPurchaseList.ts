import { calculateSatsFromEur } from "./calculateSatsFromEur";
import {
  ExchangeRateType,
  IBuildPurchaseParams,
  IPurchase,
} from "./interfaces";

export async function buildPurchaseList(
  params: IBuildPurchaseParams
): Promise<IPurchase[]> {
  const btcUsdRates = await params.exchangeRateProvider(
    params.requestList.map((r) => r.date),
    ExchangeRateType.BtcToUsd
  );
  const usdEurRates = await params.exchangeRateProvider(
    params.requestList.map((r) => r.date),
    ExchangeRateType.UsdToEur
  );

  const purchases: IPurchase[] = params.requestList.map((request, index) => ({
    date: request.date,
    amountInEur: request.amountInEur,
    btcUsdRate: btcUsdRates[index],
    usdEurRate: usdEurRates[index],
    acquiredSatoshis: calculateSatsFromEur({
      amountInEur: request.amountInEur,
      btcUsdRate: btcUsdRates[index],
      usdEurRate: usdEurRates[index],
    }),
  }));

  return purchases;
}
