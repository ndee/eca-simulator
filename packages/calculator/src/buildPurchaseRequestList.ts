import moment from "moment";
import { IEcaPurchase, IPurchaseRequest, PurchaseRepeat } from "./interfaces";

export function buildPurchaseRequestList(
  ecaPurchase: IEcaPurchase
): IPurchaseRequest[] {
  if (ecaPurchase.purchaseStart > ecaPurchase.purchaseEnd) {
    throw new Error(
      "Purchase start date cannot be later than purchase end date"
    );
  }
  const requests: IPurchaseRequest[] = [];
  let purchaseMoment = moment(ecaPurchase.purchaseStart);

  while (purchaseMoment.isBefore(moment(ecaPurchase.purchaseEnd))) {
    let momentUnit: moment.unitOfTime.DurationConstructor;
    switch (ecaPurchase.purchaseRepeat) {
      case PurchaseRepeat.Yearly:
        momentUnit = "year";
        break;
      case PurchaseRepeat.Monthly:
        momentUnit = "month";
        break;
      case PurchaseRepeat.Weekly:
        momentUnit = "week";
        break;
      case PurchaseRepeat.Daily:
        momentUnit = "day";
        break;
    }
    purchaseMoment.add(1, momentUnit);
    requests.push({
      amountInEur: ecaPurchase.purchaseAmountInEur,
      date: purchaseMoment.toDate(),
    });
  }

  return requests;
}
