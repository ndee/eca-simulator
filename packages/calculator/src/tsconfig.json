{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "outDir": "../dist",
    "composite": true,
    "declaration": true,
    "skipLibCheck": true
  },
  "include": ["**/*", "../test/exchangeRateProvider/getBtcRateFromTimeChainCalendar.spec.ts"]
}
