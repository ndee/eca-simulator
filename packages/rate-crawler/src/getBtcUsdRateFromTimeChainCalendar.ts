import axios from "axios";

export const getBtcUsdRateFromTimeChainCalendar = async (date: Date) => {
  const timestamp = date.getTime();
  const blocksAroundTimestampResponse = await axios.get(
    `https://corsproxy.io/?https%3A%2F%2Fblockchain.info%2Fblocks%2F${timestamp}%3Fformat%3Djson`
  );
  const blockHeight = blocksAroundTimestampResponse.data[0].height;
  const priceResponse = await axios.get(
    `https://tcc.qrsnap.io/block/${blockHeight}/price`
  );
  const price = priceResponse.data.result.price;
  return price;
};
