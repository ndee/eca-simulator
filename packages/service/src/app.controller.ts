import { calculateAcquiredBitcoin } from '@bg-eca-simulator/calculator';
import { IEcaPurchase } from '@bg-eca-simulator/calculator/dist/interfaces';
import { Controller, Get } from '@nestjs/common';
import { AppService } from './app.service';

@Controller()
export class AppController {
  constructor(private readonly appService: AppService) {}

  @Get()
  async getHello(): Promise<string> {
    return JSON.stringify(await calculateAcquiredBitcoin({} as IEcaPurchase));
  }
}
