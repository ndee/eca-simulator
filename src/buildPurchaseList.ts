import { calculateSatsFromEur } from "./calculateSatsFromEur";
import { IBuildPurchaseParams, IPurchase } from "./interfaces";

export async function buildPurchaseList(
  params: IBuildPurchaseParams
): Promise<IPurchase[]> {
  const purchases: IPurchase[] = [];

  for (let request of params.requestList) {
    const exchangeRateBtcUsd = await params.getBtcUsdExchangeRate(request.date);
    const exchangeRateUsdEur = await params.getUsdEurExchangeRate(request.date);
    purchases.push({
      date: request.date,
      amountInEur: request.amountInEur,
      exchangeRateBtcUsd,
      exchangeRateUsdEur,
      acquiredSatoshis: calculateSatsFromEur({
        amountInEur: request.amountInEur,
        btcUsdRate: exchangeRateBtcUsd,
        exchangeRateUsdEur,
      }),
    });
  }

  return purchases;
}
