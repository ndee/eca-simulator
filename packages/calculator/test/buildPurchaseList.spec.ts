import { buildPurchaseList } from "../src/buildPurchaseList";
import { ExchangeRateProvider, IPurchaseRequest } from "../src/interfaces";

describe("buildPurchaseList", () => {
  let rateProviderMock: ExchangeRateProvider;
  beforeEach(() => {
    rateProviderMock = jest.fn().mockResolvedValue([1, 2, 3, 4, 5, 6]);
  });
  it("should run", async () => {
    const requests: IPurchaseRequest[] = [
      {
        amountInEur: 100,
        date: new Date("2020-01-01"),
      },
      {
        amountInEur: 100,
        date: new Date("2021-01-01"),
      },
      {
        amountInEur: 100,
        date: new Date("2022-01-01"),
      },
      {
        amountInEur: 100,
        date: new Date("2023-01-01"),
      },
    ];
    expect(
      await buildPurchaseList({
        requestList: requests,
        exchangeRateProvider: rateProviderMock,
      })
    ).toMatchSnapshot();
  });
});
