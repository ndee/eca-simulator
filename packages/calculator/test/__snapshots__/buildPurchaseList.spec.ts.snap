// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildPurchaseList should run 1`] = `
[
  {
    "acquiredSatoshis": 10000000000,
    "amountInEur": 100,
    "btcUsdRate": 1,
    "date": 2020-01-01T00:00:00.000Z,
    "usdEurRate": 1,
  },
  {
    "acquiredSatoshis": 2500000000,
    "amountInEur": 100,
    "btcUsdRate": 2,
    "date": 2021-01-01T00:00:00.000Z,
    "usdEurRate": 2,
  },
  {
    "acquiredSatoshis": 1111111111,
    "amountInEur": 100,
    "btcUsdRate": 3,
    "date": 2022-01-01T00:00:00.000Z,
    "usdEurRate": 3,
  },
  {
    "acquiredSatoshis": 625000000,
    "amountInEur": 100,
    "btcUsdRate": 4,
    "date": 2023-01-01T00:00:00.000Z,
    "usdEurRate": 4,
  },
]
`;
